### DISCUSSION TITLE
App instantiation

### DISCUSSION THREAD

**Discussion by meirumeiru** [microsoft/WindowsAppSDK#2251] [Q&A]
I'm trying to build an application that uses Xaml islands with WinUI3. The problem I have with this is, that I need to instantiate an "Application" object.

1) Why do I need it? Does it have to have it's own thread or something? When it runs... is it something like a "global" thing that every window then uses to link to? ... e.g. like... "registered into the thread local variables??" ... I've no idea how all this works together and how the "magic" finding of the correct objects works in the background...

2) How do I do this? ...

a)
winrt::MyApp::App app;

b)
winrt::Microsoft::UI::Xaml::Application::Start(
      [](auto &&)
      { winrt::make<MyApp>(); });

c)
auto app = winrt::make<MyApp>();

... why does e.g. c) throw an error E_RPC_WRONG_THREAD ??

Sorry, didn't find enough information to find out what it does and how those classes work. It's hard to follow those massive amounts of template inheritance trees...

Rudolf

**Comment by castorix** [comment]
XAML Island for WinUI3 is not yet implemented :  [under-consideration](https://portal.productboard.com/winappsdk/1-windows-app-sdk/tabs/1-under-consideration)
As I wrote in [this thread](https://github.com/microsoft/WindowsAppSDK/discussions/2233), you must use _MddBootstrapInitialize_  to register XAML controls, then you can instantiate Application + Window
To make it working, I copied parts of the code from a C++/WinRT WinUI3 Template, like _App.xaml.g.h_

**Reply by meirumeiru** [reply]
ok... but, can you explain a little bit, how an Application-object has to be built? a, b or c? ... does it have its own thread? and what is this E_RPC_WRONG_THREAD error? ... my problem is, that all the documents I find, they are some sort of "good weather"-documents... you learn nothing about how it really works and then when something doesn't work, you have no idea what to do

if I could get some background information (or, tell me if there's a video or blog or whatever) about those architectures (maybe about the old WPF? or WinUI2? if it is similar)... but currently all I see are compiletime or runtime errors of which I have no idea where they could come from

and I think it is mainly, because I've no idea how the "Application"-class works

**Reply by castorix** [reply]
In my test, I used (_winrt::WinRTTest2::implementation_ is the namespace) :

```
	winrt::init_apartment(winrt::apartment_type::single_threaded);	
	winrt::Microsoft::UI::Xaml::Application::Start(
		[](auto&&)
		{
			::winrt::make<::winrt::WinRTTest2::implementation::App>();
		});
```

**Reply by meirumeiru** [reply]
> In my test, I used (_winrt::WinRTTest2::implementation_ is the namespace) :
> 	winrt::Microsoft::UI::Xaml::Application::Start ...

*hmm* ok... I think I start to understand how it works... the Application is a Dispatcher and the Dispatcher is running a message loop (at least that's what I think) ... and somehow it's managing, that the correct callbacks/handlers are called. No longer via a "window proc", but in a way still similar... and, I guess that not everything goes through this message loop (or at least much less than in the win32 world) ...
it also seems that the idea of "SendMessage executes directly when it's in the same thread or blocks until the loop of the other thread has processed the massage" has been replaced with a "throw an exception if you're not calling from the correct thread or... use the dispatcher yourself"...
fine... and to me it seems that the "Application::Start" does somehow run this message loop... and therefor I think, you cannot use this in a Xaml-Island example, because here we have to use an external loop... (so, no idea how you got this to work the way you say you did it)

this needs a lot more digging into it until I fully know how it's working and until I know how to use it

... and, sorry, but I still have problems reading those codes that use templates that heavily. But, that's normal today... I know... templates and partial classes. With open parameter lists that are never used for more or less than exactly 1 parameter... but that's how it's done today :-) ... what a mess

**Reply by castorix** [reply]
> sorry, but I still have problems reading those codes that use templates that heavily.

Yes, me too... although I program in C/C++ since 80's, but I don't like the WinRT syntax and I prefer C# for those things, a lot simpler and easier to understand (like _public partial class App : Application_)
